### PR TITLE
gather: update the state handling for terraform 0.12

### PR DESCRIPTION
**Pull request by abhinavdahiya**
- pkg/terraform: utilize the vendored terraform to read tfstate

    Utilizing the vendored statefile serialize/deserialize helps installer to no worry about the statefile
    version conversions. Installer can always assume to use the latest/chosen version of state file from vendored code.

    For example, terraform 0.12 moves the terraform state to version 4, compared to version 3. With this change installer can assume the terraform state
    deserialized for use will always be version 4, as it will be upconverted by vendored terraform statefile package.

- pkg/terraform: add gather package to store platform specific code from terraform

    This adds bootstrap and control plane hosts ip gather from state file for platforms.

- cmd/openshift-install: update gather code to utilize new terraform packages

/cc @wking @jstuever

### CODE REVIEW

**Comment by jstuever** [comment]
Looks solid to me. I'll defer lgtm to Trevor.

File: pkg/terraform/exec/state.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"bytes"
+	"os"
+
+	"github.com/hashicorp/terraform/states/statefile"
+	"github.com/pkg/errors"
+)
+
+// ReadState reads the terraform state from file and returns the contents in bytes
+// It returns an error if reading the state was unsuccessful
+// RadState utilizes the terraform's internal wiring to upconvert versions of terraform state to return
```
Comment (rgolangh) [inline pkg/terraform/exec/state.go]: RadState ReadState

Comment (abhinavdahiya) [reply pkg/terraform/exec/state.go]: done with https://github.com/openshift/installer/pull/1763/commits/a9fb11401e898113e511d9d36a3e0c3366be6e22 -> https://github.com/openshift/installer/pull/1763/commits/0399139ea4f5bf85285a0c2c8547c0d5d2be9033

**Comment by abhinavdahiya** [comment]
friendly ping @jstuever and @wking

**Comment by wking** [comment]
> For example, terraform 0.12 moves the terraform state to version 4, compared to version 3. With this change installer can assume the terraform state deserialized for use will always be version 4, as it will be upconverted by vendored terraform statefile package.

Do we need this?  Running two different installer versions on the same Terraform state file seems like a bigger problem than just "what version is the state file?".  For example, maybe we've shuffled some modules around and whatever you were trying to access (the bootstrap IP address?) is now in a different place?

**Comment by abhinavdahiya** [comment]
> > For example, terraform 0.12 moves the terraform state to version 4, compared to version 3. With this change installer can assume the terraform state deserialized for use will always be version 4, as it will be upconverted by vendored terraform statefile package.
> 
> Do we need this? Running two different installer versions on the same Terraform state file seems like a bigger problem than just "what version is the state file?". 

Yes, we need to try our best to not break people. Keep things backward compatible. This makes sure we can at-least read the terraform state file.

> For example, maybe we've shuffled some modules around and whatever you were trying to access (the bootstrap IP address?) is now in a different place?

In that case, we'll need to make sure we makes changes to gather to understand old and new locations. If we can keep backward compatible we should.

**Comment by wking** [comment]
> Yes, we need to try our best to not break people. Keep things backward compatible. This makes sure we can at-least read the terraform state file.

But why would you want to read the Terraform state file written by a different installer?  The only story I can think of is:

1. User uses installer-A attempts to create a cluster and fails in Terraform.
2. User uses installer-B to `openshift-install gather bootstrap`.

But that's a pretty short window in which to be switching installers.  And other actions like changing installers between an `openshift-install create manifests` and an `openshift-install create cluster` seem even riskier.  I'd rather have the only compat contract between different installer versions reading eachother's assets be `install-config.yaml` (multi-cluster lifespan) and `metadata.json` (full cluster lifespan).  The other assets, and especially the Terraform state, seem like "cluster-install lifespan" where requiring a consistent installer version is not a hardship.

**Comment by abhinavdahiya** [comment]
> > Yes, we need to try our best to not break people. Keep things backward compatible. This makes sure we can at-least read the terraform state file.
> 
> But why would you want to read the Terraform state file written by a different installer? The only story I can think of is:
> 
> 1. User uses installer-A attempts to create a cluster and fails in Terraform.
> 2. User uses installer-B to `openshift-install gather bootstrap`.
> 
> But that's a pretty short window in which to be switching installers. And other actions like changing installers between an `openshift-install create manifests` and an `openshift-install create cluster` seem even riskier. I'd rather have the only compat contract between different installer versions reading eachother's assets be `install-config.yaml` (multi-cluster lifespan) and `metadata.json` (full cluster lifespan). The other assets, and especially the Terraform state, seem like "cluster-install lifespan" where requiring a consistent installer version is not a hardship.

I'm not sure why the point of contention is depending on terraform code itself to read the state file.
The reasoning is it's one less thing Installer needs to think about in terms of maintaining, Installer shouldn't need to care what version was written at, it cares I know and expect it to look like this.

**Pull request merged**